src/DoubleSpendProof.cpp

```diff
@@ -0,0 +1,320 @@
+        diff = m_spender1.hashPrevOutputs.Compare(m_spender2.hashPrevOutputs);
+    if (diff > 0)
+        return Invalid;
+
+    // Get the previous output we are spending.
+    int64_t amount;
+    CScript prevOutScript;
+    auto prevTx = mempool.get(m_prevTxId);
+    if (prevTx.get())
+    {
+        if (prevTx->vout.size() <= (size_t)m_prevOutIndex)
+            return Invalid;
+        auto output = prevTx->vout.at(m_prevOutIndex);
+        amount = output.nValue;
+        prevOutScript = output.scriptPubKey;
+    }
+    else
+    {
+        Coin coin;
+        if (!pcoinsTip->GetCoin({m_prevTxId, (uint32_t)m_prevOutIndex}, coin))
+        {
+            /* if the output we spend is missing then either the tx just got mined
+             * or, more likely, our mempool just doesn't have it.
+             */
+            return MissingUTXO;
+        }
+        amount = coin.out.nValue;
+        prevOutScript = coin.out.scriptPubKey;
+    }
+
+    /*
+     * Find the matching transaction spending this. Possibly identical to one
+     * of the sides of this DSP.
+     * We need this because we want the public key that it contains.
+     */
+    CTransaction tx;
+    {
+        READLOCK(mempool.cs_txmempool);
+        auto it = mempool.mapNextTx.find({m_prevTxId, (uint32_t)m_prevOutIndex});
+        if (it == mempool.mapNextTx.end() || (size_t)m_prevOutIndex >= it->second.ptx->vout.size())
+        {
+            return MissingTransaction;
+        }
+        tx = *(it->second.ptx);
+    }
+
+    /*
+     * TomZ: At this point (2019-07) we only support P2PKH payments.
+     *
+     * Since we have an actually spending tx, we could trivially support various other
+     * types of scripts because all we need to do is replace the signature from our 'tx'
+     * with the one that comes from the DSP.
+     */
+    ScriptType scriptType = P2PKH; // FUTURE: look at prevTx to find out script-type
+
+    std::vector<uint8_t> pubkey;
+    for (size_t i = 0; i < tx.vin.size(); ++i)
+    {
+        if (tx.vin[i].prevout.n == (size_t)m_prevOutIndex && tx.vin[i].prevout.hash == m_prevTxId)
+        {
+            // Found the input script we need!
+
+            CScript inScript = tx.vin[i].scriptSig;
+            auto scriptIter = inScript.begin();
+            opcodetype type;
+            inScript.GetOp(scriptIter, type); // P2PKH: first signature
+            inScript.GetOp(scriptIter, type, pubkey); // then pubkey
+            break;
+        }
+    }
+
+    if (pubkey.empty())
+        return Invalid;
+
+    CScript inScript;
+    if (scriptType == P2PKH)
+    {
+        inScript << m_spender1.pushData.front();
+        inScript << pubkey;
+    }
+    DSPSignatureChecker checker1(this, m_spender1, amount);
+    ScriptError_t error;
+    if (!VerifyScript(inScript, prevOutScript, 0 /*flags*/, MAX_OPS_PER_SCRIPT, checker1, &error))
+    {
+        LOGA("DoubleSpendProof failed validating first tx due to %s\n", ScriptErrorString(error));
```

Comment:
Should use `LOG(DSPROOF, ...)` instead of `LOGA()`

Reply:
done